### PR TITLE
feat: add iter/while-each

### LINKED ISSUE #810: [RFC]: Add `@stdlib/iter/while-each`
### Description

This RFC proposes adding an iterator utility to invoke a function for each iterated value until a predicate function returns `false`.

This is the iterator equivalent of [`@stdlib/utils/while-each`](https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/utils/while-each).

Package: `@stdlib/iter/while-each`
Alias: `iterWhileEach`

### Related Issues

None.

### Questions

No.

### Other

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by Jaysukh-409**
Resolves #810.

## Description

> What is the purpose of this pull request?

This pull request:

-   Create an iterator which, while a test condition is true, invokes a function for each iterated value before returning the iterated value.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #810 

## Questions

> Any questions for reviewers of this pull request?

I have added lib and examples for the whileEach iterator. Please review and provide feedback accordingly.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+-->
+
+# whileEach Iterator
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
# iterWhileEach
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+## Usage
+
+```javascript
+var iterWhileEach = require( '@stdlib/iter/while-each' );
+```
+
+#### iterWhileEach( iterator, predicate, fcn\[, thisArg] )
+
+Returns an iterator which invokes a `function` for each iterated value untill either a `predicate` returns `false` or the function has iterated over all value **before** returning the iterated value.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
Returns an iterator which invokes a function for each iterated value **before** returning the iterated value until either a `predicate` function returns `false` or the iterator has iterated over all values.
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+r = it.next().value;
+// undefined
+
+// ...
+```
+
+The returned iterator protocol-compliant object has the following properties:
+
+-   **next**: function which returns an iterator protocol-compliant object containing the next iterated value (if one exists) assigned to a `value` property and a `done` property having a `boolean` value indicating whether the iterator is finished.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
-   **next**: function which returns an iterator protocol-compliant object containing the next iterated value (if one exists) assigned to a `value` property and a `done` property having a boolean value indicating whether the iterator is finished.
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+-   **return**: function which closes an iterator and returns a single (optional) argument in an iterator protocol-compliant object.
+
+The invoked `function` is provided two arguments:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
Both the `predicate` function and the function to invoke for each iterated value are provided two arguments:
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+-   `value`: iterated value
+-   `index`: iteration index (zero-based)
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
-   **value**: iterated value
-   **index**: iteration index (zero-based)
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+
+function predicate( v ) {
+    return v < 3;
+}
+
+function assert( v ) {
+    if ( v !== v ) {
+        throw new Error( 'should not be NaN' );
+    }
+}
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
function assert( v, i ) {
    if ( i > 1 ) {
        throw new Error( 'unexpected error' );
    }
}
```

I see that `iter/for-each` has a similar example, which does not actually illustrate anything new. Here, we at least want to show that provided functions are invoked with multiple arguments.

Comment (kgryte) [reply lib/node_modules/@stdlib/iter/while-each/README.md]: Once this PR is merged, feel free to submit a PR against `iter/for-each` which updates the respective example to be more illustrative.

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+r = it.next().value;
+// returns 2
+
+r = it.next().value;
+// undefined
+
+// ...
+```
+
+To set the function execution context, provide a `thisArg`.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
To set the execution context for `fcn`, provide a `thisArg`.
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+
+```javascript
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md:117]:
```suggestion
<!-- eslint-disable no-undefined-this -->

```javascript
```

I believe we need to disable this lint rule here, as I believe we should be getting a lint error.

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+<section class="examples">
+
+## Examples
+
+<!-- eslint no-undef: "error" -->
+
+```javascript
+var randu = require( '@stdlib/random/iter/randu' );
+var isnan = require( '@stdlib/math/base/assert/is-nan' );
+var iterForEach = require( '@stdlib/iter/for-each' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
var iterWhileEach = require( '@stdlib/iter/while-each' );
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+}
+
+// Create a seeded iterator for generating pseudorandom numbers:
+var rand = randu({
+    'seed': 1234,
+    'iter': 10
+});
+
+// Create an iterator which validates generated numbers:
+var it = iterForEach( rand, predicate, assert );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
var it = iterWhileEach( rand, predicate, assert );
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/iter/map`][@stdlib/iter/map]</span><span class="delimiter">: </span><span class="description">create an iterator which invokes a function for each iterated value.</span>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
```

We handle "See Also" sections in a separate build process which draws from a central project database.

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 12]
```diff
@@ -0,0 +1,243 @@
...
+<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="links">
+
+<!-- <related-links> -->
+
+[@stdlib/iter/map]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/iter/map
+
+<!-- </related-links> -->
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/iter/while-each/docs/repl.txt [hunk 13]
```diff
@@ -0,0 +1,53 @@
+
+{{alias}}( iterator, predicate, fcn[, thisArg] )
+    Returns an iterator which invokes a function for each iterated value before
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]: This description needs to be updated to match the README.

File: lib/node_modules/@stdlib/iter/while-each/docs/repl.txt [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+    returning the iterated value.
+
+    When invoked, the input function is provided two arguments:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]:
```suggestion
    When invoked, both input functions are provided two arguments:
```

File: lib/node_modules/@stdlib/iter/while-each/docs/repl.txt [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+    If an environment supports Symbol.iterator, the returned iterator is
+    iterable.
+
+    Parameters
+    ----------
+    iterator: Object
+        Input iterator.
+
+    predicate: Function
+        Function which indicates whether to continue iterating or not.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]:
```suggestion
        Function which indicates whether to continue iterating.
```

File: lib/node_modules/@stdlib/iter/while-each/docs/repl.txt [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+        Returns an iterator protocol-compliant object containing the next
+        iterated value (if one exists) and a boolean flag indicating whether the
+        iterator is finished.
+
+    iterator.return( [value] ): Function
+        Finishes an iterator and returns a provided value.
+
+    Examples
+    --------
+    > function predicate( v ) { return v === v};
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]:
```suggestion
    > function predicate( v ) { return v === v };
```

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+var iterForEach = require( '@stdlib/iter/for-each' );
+
+function assert( v ) {
+    if ( isnan( v ) ) {
+        throw new Error( 'should not be NaN' );
+    }
+}
+
+function predicate( v ) {
+    return v < 0.5;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
    return v <= 0.75;
```

Updating to match `examples/index.js`.

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+var isFunction = require( '@stdlib/assert/is-function' );
+var isIteratorLike = require( '@stdlib/assert/is-iterator-like' );
+var iteratorSymbol = require( '@stdlib/symbol/iterator' );
+var format = require( '@stdlib/string/format' );
+
+
+// MAIN //
+
+/**
+* Returns an iterator which invokes a function for each iterated value before returning the iterated value.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]: This description needs to be updated.

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+		if ( v.done ) {
+			FLG = true;
+			return v;
+		}
+		v = v.value;
+		fcn.call( thisArg, v, i );
+		return {
+			'value': v,
+			'done': !predicate( v )
+		};
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]:
```suggestion
		if ( v.done ) {
			FLG = true;
			return v;
		}
		v = v.value;
		if ( predicate( v, i ) === false ) {
			FLG = true;
			return {
				'done': true
			};
		}
		fcn.call( thisArg, v, i );
		return {
			'value': v,
			'done': false
		};
```

As currently implemented, `fcn` is being called **before** `predicate` which is not what should happen (see [`@stdlib/utils/while-each`](https://github.com/stdlib-js/stdlib/blob/30cf48578491d545c528a28ccc031545c8e7bbf9/lib/node_modules/%40stdlib/utils/while-each/lib/main.js#L71)). We want to stop calling `fcn` as soon as `predicate` returns `false`; at which point, the iterator is considered "done".

Additionally, as stated in the docs, `predicate` should be provided both the value and index.

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+	}
+
+	/**
+	* Returns a new iterator.
+	*
+	* @private
+	* @returns {Iterator} iterator
+	*/
+	function factory() {
+		return iterWhileEach( iterator[ iteratorSymbol ](), predicate, fcn, thisArg );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]:
```suggestion
		return iterWhileEach( iterator[ iteratorSymbol ](), predicate, fcn, thisArg ); // eslint-disable-line max-len
```

**Review by kgryte — CHANGES_REQUESTED** [review]
@Jaysukh-409 Thanks for working on this! A few descriptions need to be updated, and this package still needs tests. For the latter, you can consult `iter/for-each` while borrowing some ideas from `utils/while-each`. Once resolved, this PR should be ready for merge.

File: lib/node_modules/@stdlib/iter/while-each/README.md [hunk 21]
```diff
@@ -0,0 +1,233 @@
...
+
+r = it.next().value;
+// undefined
+
+// ...
+```
+
+To set the execution context for `fcn`, provide a `thisArg`.
+
+<!-- eslint-disable no-undefined-this -->
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/README.md]:
```suggestion
<!-- eslint-disable no-invalid-this -->
```

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 22]
```diff
@@ -0,0 +1,161 @@
...
+*     return ( v <= 0.5 );
+* }
+*
+* function assert( v ) {
+*     if ( isnan( v ) ) {
+*         throw new Error( 'should not be NaN' );
+*     }
+* }
+*
+* var iter = iterWhileEach( randu(), predicate, assert );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]: @Jaysukh-409 I suggest not using `randu()` here, as the doctests (i.e., `returns <number>`) will randomly fail. I suggest using the example from the README which uses `array2iterator` and a static array `[ 1, 2, 3, 4 ]` with the `predicate` function updated accordingly.

File: lib/node_modules/@stdlib/iter/while-each/lib/index.js [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+*     return ( v <= 0.5 );
+* }
+*
+* function assert( v ) {
+*     if ( isnan( v ) ) {
+*         throw new Error( 'should not be NaN' );
+*     }
+* }
+*
+* var iter = iterWhileEach( randu(), predicate, assert );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/index.js]: @Jaysukh-409 Same comment as https://github.com/stdlib-js/stdlib/pull/1180/files#r1427553714.

File: lib/node_modules/@stdlib/iter/while-each/docs/repl.txt [hunk 24]
```diff
@@ -1,7 +1,8 @@
 
 {{alias}}( iterator, predicate, fcn[, thisArg] )
-    Returns an iterator which invokes a function for each iterated value before
-    returning the iterated value.
+    Returns an iterator which invokes a function for each iterated value before 
+    returning the iterated value until either a predicate function
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]: Just checking: do these lines wrap at 80 characters?

Comment (kgryte) [reply lib/node_modules/@stdlib/iter/while-each/docs/repl.txt]: Line 4 looks like it wraps prematurely.

File: lib/node_modules/@stdlib/iter/while-each/lib/index.js [hunk 25]
```diff
@@ -0,0 +1,63 @@
...
+* }
+*
+* var it = iterWhileEach( array2iterator( [ 1, 2, 3, 4 ] ), predicate, assert );
+* // returns {}
+*
+* var r = iter.next().value;
+* // returns <number>
+*
+* r = iter.next().value;
+* // returns <number>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/index.js]:
```suggestion
* // returns 1
*
* r = iter.next().value;
* // returns 2
```

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 26]
```diff
@@ -0,0 +1,165 @@
...
+* }
+*
+* function assert( v ) {
+*     if ( v !== v ) {
+*         throw new Error( 'should not be NaN' );
+*     }
+* }
+*
+* var it = iterWhileEach( array2iterator( [ 1, 2, 3, 4 ] ), predicate, assert );
+* // throws <ReferenceError>
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]: @Planeshifter Looks like we are still getting ESLint auto-fixes in doctest comments. Would you mind ensuring that this gets resolved?

@Jaysukh-409 In the meantime, it looks like we have error. Namely, Lines 45-46 are incorrect. You need to import `array2iterator`.

File: lib/node_modules/@stdlib/iter/while-each/lib/index.js [hunk 27]
```diff
@@ -0,0 +1,63 @@
...
+'use strict';
+
+/**
+* Create an iterator which, while a test condition is true, invokes a function for each iterated value before returning the iterated value.
+*
+* @module @stdlib/iter/while-each
+*
+* @example
+* var randu = require( '@stdlib/random/iter/randu' );
+* var isnan = require( '@stdlib/math/base/assert/is-nan' );
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/index.js]: These imports are no longer necessary. And we need to import `array2iterator`.

**Comment by Jaysukh-409** [comment]
Please ignore changes made above. Due to some issues I got this trouble.

File: lib/node_modules/@stdlib/iter/while-each/lib/main.js [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+var isFunction = require( '@stdlib/assert/is-function' );
+var isIteratorLike = require( '@stdlib/assert/is-iterator-like' );
+var iteratorSymbol = require( '@stdlib/symbol/iterator' );
+var format = require( '@stdlib/string/format' );
+
+
+// MAIN //
+
+/**
+* Returns an iterator which invokes a function for each iterated value before returning the iterated value until either a predicate function returns false or the iterator has iterated over all values.
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/lib/main.js]:
```suggestion
* Returns an iterator which invokes a function for each iterated value before returning the iterated value until either a predicate function returns `false` or the iterator has iterated over all values.
```

File: lib/node_modules/@stdlib/iter/while-each/test/test.js [hunk 29]
```diff
@@ -0,0 +1,368 @@
...
+	var i;
+
+	it = iterWhileEach( randu(), predicate, assert );
+	t.equal( it.next.length, 0, 'has zero arity' );
+
+	count = 0;
+	i = 0;
+	do {
+		r = it.next();
+		if ( r.value ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/test/test.js]:
```suggestion
		if ( typeof r.value !== 'undefined' ) {
```

`r.value` can be `0`.

File: lib/node_modules/@stdlib/iter/while-each/test/test.js [hunk 30]
```diff
@@ -0,0 +1,368 @@
...
+	};
+	rand = randu( opts );
+	it = iterWhileEach( rand, predicate, assert );
+	t.equal( it.next.length, 0, 'has zero arity' );
+
+	expected = [];
+	i = 0;
+	do {
+		r = it.next();
+		if ( r.value ) {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/test/test.js]:
```suggestion
		if ( typeof r.value !== 'undefined' ) {
```

`r.value` can be `0`.

**Review by kgryte — CHANGES_REQUESTED** [review]
@Jaysukh-409 This is looking good! I think the only thing missing is benchmarks.

**Comment by kgryte** [comment]
...and TypeScript declarations.

File: lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts [hunk 31]
```diff
@@ -0,0 +1,144 @@
...
+* @returns a boolean indicating whether to continue iterating or not
+*/
+type Predicate = nullaryPredicate | unaryPredicate | binaryPredicate;
+
+/**
+* Returns an iterator which invokes a function for each iterated value **before** returning the iterated value until either a predicate function returns `false` or the iterator has iterated over all values.
+*
+* ## Notes
+*
+* -   When invoked, both callback and `predicate` function are provided two arguments:
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts]:
```suggestion
* -   When invoked, both the `predicate` and callback functions are provided two arguments:
```

File: lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts [hunk 32]
```diff
@@ -0,0 +1,144 @@
...
+*
+* r = it.next().value;
+* // returns 2
+*
+* r = it.next().value;
+* // undefined
+*
+* // ...
+*/
+declare function iterWhileEach( iterator: Iterator, predicate: Predicate ,fcn: Callback, thisArg?: any ): Iterator;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts]:
```suggestion
declare function iterWhileEach( iterator: Iterator, predicate: Predicate, fcn: Callback, thisArg?: any ): Iterator;
```

File: lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+*
+* r = it.next().value;
+* // returns 2
+*
+* r = it.next().value;
+* // undefined
+*
+* // ...
+*/
+declare function iterWhileEach( iterator: Iterator, predicate: Predicate, fcn: Callback, thisArg?: any ): Iterator;
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/types/index.d.ts:139]: We can improve this type declaration, but we need to do this more generally for packages in `iter/*`, where we are not good about preserving type information. We can leave as is, for now, and address in a subsequent PR/commit.

File: lib/node_modules/@stdlib/iter/while-each/docs/types/test.ts [hunk 34]
```diff
@@ -0,0 +1,121 @@
...
+
+import iterWhileEach = require( './index' );
+
+/**
+* Returns an iterator protocol-compliant object.
+*
+* @returns iterator protocol-compliant object
+*/
+function iterator(): any {
+
```
Comment (kgryte) [inline lib/node_modules/@stdlib/iter/while-each/docs/types/test.ts]:
```suggestion
```

**Review by kgryte — APPROVED** [review]
LGTM. Thanks, @Jaysukh-409!

**Pull request merged**